examples/maybe_error.cpp

```diff
@@ -0,0 +1,62 @@
+// Copyright 2022 PickNik Inc
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//    * Redistributions of source code must retain the above copyright
+//      notice, this list of conditions and the following disclaimer.
+//
+//    * Redistributions in binary form must reproduce the above copyright
+//      notice, this list of conditions and the following disclaimer in the
+//      documentation and/or other materials provided with the distribution.
+//
+//    * Neither the name of the PickNik Inc nor the names of its
+//      contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <monad/all.hpp>
+
+namespace views = ::ranges::views;
+
+Result<double> divide_4_by(double x) {
+  if (x == 0.0) {
+    return make_unexpected(InvalidArgument(fmt::format("divide by 0")));
+  }
+  return 4.0 / x;
+}
+
+Result<double> f1(double x) { return divide_4_by(x); }
+Result<double> f2(double x) { return divide_4_by(x); }
+Result<double> f3(double x) { return divide_4_by(x); }
+Result<double> f4(double x) { return divide_4_by(x); }
+
+Result<double> do_work() {
+  auto const a = make_result(2.4) | f1 | f2;
+  auto const b = make_result(0.0) | f3 | f4;
+
+  if (auto const error = maybe_error(a, b); error) {
```

Comment:
It really *feels* like this should return a `Result`, and not a `std::optional`, then couldn't we write
```
if (auto const error = maybe_error(a, b); error) {
  return error;
}
```
or am I missing something?

Reply:
If this returned a Result<T> what would the T be?  and would it even be useful.  The original idea was that T would be a tuple of all the values... but I'm not sure that's more useful than an optional<E> as what you want to test for is if any of them failed and return that error.

Reply:
So there's no `Result` type like there is a `Status` type that isn't `StatusOr`?

Reply:
no... I was sort of copying the rust names with

Result<T> = expected<T, Error>;